clippy_lints/src/question_mark.rs

```diff
@@ -122,8 +153,56 @@ impl QuestionMark {
         }
     }
 
-    fn result_check_and_early_return(cx: &LateContext<'_>, expr: &Expr<'_>, nested_expr: &Expr<'_>) -> bool {
-        Self::is_result(cx, expr) && Self::expression_returns_unmodified_err(cx, nested_expr, expr)
+    /// Check whether the encountered expression contains `else if`
+    fn expr_is_else_if(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+        if_chain! {
+            if let ExprKind::If(..) = expr.kind;
+            if let Some(parent_expr) = clippy_utils::get_parent_expr(cx, expr);
+            if let ExprKind::If(..) = parent_expr.kind;
+            then {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Check whether the fixed code needs semicolon after `?`
+    ///
+    /// The fixed code of this expression doesn't need semi
+    /// ```ignore
+    /// let _ = if let Some(x) = Some(0) {
+    ///     x
+    /// } else {
+    ///     return None;
+    /// };
+    /// ```
+    ///
+    /// The fixed code of this expression needs semi
+    /// ```ignore
+    /// if let Err(err) = foo() {
+    ///     return Err(err);
+    /// }
+    /// ```
+    fn is_semicolon_needed(if_then: &Expr<'_>, if_else: Option<&'_ Expr<'_>>) -> bool {
+        let if_then_kind = &peel_blocks_with_stmt(if_then).kind;
+        let if_then_is_ret_stmt = matches!(if_then_kind, ExprKind::Ret(..));
+
+        if if_else.is_none() {
+            return if_then_is_ret_stmt;
+        }
+        if let ExprKind::Ret(_) = peel_blocks_with_stmt(if_else.unwrap()).kind {
+            return if_then_is_ret_stmt;
+        }
+        false
+    }
```

Comment:
I would expect this to be implemented by checking whether the parent node of the expression is a statement.

Reply:
Did you suggesting `clippy_utils::get_parent_node`? if yes, I don't understand how am I going to do with it, since the parent node of them will give me the entire if expression. Or perhaps I checked the wrong children...

Reply:
I'm not sure what you mean. Wouldn't this work?

```rust
let requires_semi = matches!(get_parent_node(expr), Some(Node::Stmt(_)));

Reply:
Yep, it works, I was just being dumb thinking you want me to check the parent node of the return value, like the `if_then_is_ret_stmt` in this case, sorry for causing confusion.